### ISSUE TITLE
Stop ALL Tracks on a specific Output OR Allow Query

### ISSUE THREAD

**Issue by djrobby** [robertsonics/Tsunami-Arduino-Serial-Library#9]
I have `560` mono tracks that are played/stopped via a REST API communicating with an application specific NodeJS implementation of this serial protocol. Everything works as expected but I believe that the routine I have for stopping whatever `TRACK` is playing on a specific `OUTPUT` is very inefficient. I say this because I am sending `560x STOP` commands (1 per track) to any given `OUTPUT` that I want to stop from playing whatever track(s) it is currently playing. As it stands, the serial protocol only allows you to either `Stop ALL Tracks on ALL Outputs` or `Stop One (1) Track on One (1) specific Output`. I did try to maintain a state of which tracks might be playing on a specific output but, due to edge cases, it kept going out of sync with what was actually playing on Tsunami.

With that in mind, are you able to implement the ability to either:

- Send a serial command to `STOP ALL TRACKS` playing on ONE (1) specific `OUTPUT` or,
- Send a serial command to `QUERY ONE(1)` specific `OUTPUT` that returns which `TRACK(S)` it's playing

Thank you in advance [🙏](https://emojipedia.org/folded-hands/)

**Comment by robertsonics** [comment]
This is certainly a reasonable request. I can implement a command to stop all tracks on a specified output, but it will necessitate a firmware update as well. Are you using the original Tsunami or the new Tsunami Qwiic version?

**Comment by djrobby** [comment]
Thank you for your prompt response. I'm using Tsunami original for development/testing but was planning on ordering additional units for deployment. I'm assuming the original version has been phased out?

**Comment by robertsonics** [comment]
Yes - the AKM codec went end-of-life, necessitating a new version. You should also be aware that even the new version is becoming harder and harder for Sparkfun to keep in stock due to numerous parts being affected by the global supply chain issues. They've been building small batches when they can, but these sell out quickly. 

I'm actually working on a new 10-channel output player that should be easier to manufacture and keep in stock that I may sell directly from my website, but that's at least a few months off.

Give me a few days to see about implementing this new command, and I'll post back here.

**Comment by djrobby** [comment]
I understand. Supply chain is a big mess right now! I won't need production units until fall this year (Sept-Oct). As for quantities, my current forecast is 40-60pcs.

My application uses all 8-channels so this board has always been on my radar. In the past, we wrote a custom ASIO driver using [BASSASIO](https://www.un4seen.com/) audio library to drive these channels on a regular PC equipped with 4x physical audio outputs capable of 7.1 playback. However, as PCs become smaller in form factor, manufacturers have started phasing out the additional audio jacks for a single headphone output along with optical. Also, with this setup, I was limited to the Windows platform only. So, I finally decided to move off Windows and over to a browser based solution using your board. So far, it's been rock solid!

Thank you again! [🙏](https://emojipedia.org/folded-hands/)

**Comment by robertsonics** [comment]
If I may ask, have you had any problem with the unbalanced outputs and noise? I know that a small number of users have had issues with certain combinations of power supplies and external amplifiers. I've been considering balanced outputs on the new design but this would add significant cost and size.

**Comment by djrobby** [comment]
No, I have not. Attached is my current setup going 16P out from your board to 16P I/O block to a 4CH amp (production setup will use 2x amps). My application does not require high-fidelity audio as it's only driving mono tones via siren speakers. I'll be on the lookout if I hear anything and report back to you.

![Tsunami Original Setup](https://i.imgur.com/DrAQc53l.jpg)

**Comment by djrobby** [comment]
So, I have an interesting issue after letting the 4x mono tracks playback (in loop mode) on 4 independent outputs (1 track per output) for more than 3-4mins. The playback turns into a clicking noise. I have a video documenting this issue here: [https://youtu.be/FoR8l_J6mVw](https://youtu.be/FoR8l_J6mVw). Any ideas what could be causing this?

**Comment by robertsonics** [comment]
The first step is to determine if the processor on the Tsunami is still running or has gone off the rails - which is what it sounds like. Once it's in this state, does it respond to any serial communications? Can you send the stop commands and observe the LED going back to the flashing heartbeat? If not, that's evidence that the processor has crashed and the sound you're hearing is simply the audio buffer being played but not being updated.

I am rather confident after all this time and all the units out there that there are no firmware bugs that cause the processor to crash - I routinely leave a Tsunami looping for days - so the only thing left is hardware. This could be anything from a power supply issue, static discharge or a defective Tsunami. For example, a bad solder connection on the microSD socket can cause something like this, and I've seen units come from SparkFun with loose pins.

How are you powering the unit?

**Comment by djrobby** [comment]
I'm driving the Tsunami board and a 4-channel x 65W/channel amplifier using a MeanWell 12VDC 350W power supply. I just tested now and the processor does respond to the `STOP` and `STOP-ALL` commands. It even accepts the play command but after the issue starts, it continues to play the clicking noise until `RESET`. The solder joints look fine to me...I've upload some pix here for your reference: https://imgur.com/a/gt6ddnd. Please let me know and I post more pictures as needed. Thanks!

**Comment by robertsonics** [comment]
When you send the STOP-ALL, does the LED go out and resume the heartbeat flashing? Does the LED go on solid again when you issue another play command? If so, it could be a hardware issue between the CPU and the audio codec, something that's causing the codec to either reset or lose sync with the TDM protocol. If it were me here, I'd examine all the solder connections on the CPU and the codec under a microscope.

The next step would be to remove the Tsunami from your setup, use a standard 5V USB-C power supply and headphones, so that the only things connected to Tsunami are the USB power supply and the headphones, and program a trigger to play a looping sound. If it fails during this test, then it's definitely a defective unit.

The next step would be to try another Tsunami. If you only have one Tsunami, then you may have to contact SparkFun to see about a replacement.

**Comment by robertsonics** [comment]
Also, it's probably obvious, but have you tried a different microSD card? Also, is the card formatted with 32k allocation size?

**Comment by djrobby** [comment]
I just tried with 4x completely different files that I had been originally testing with and so far it's been running for 30mins just fine. I was initially testing tracks 1-4 routed to outputs 1-4 (respectively) and from what I tell, one of those tracks seems to be throwing the processor into some kind of limbo. I have uploaded METADATA from one of the tracks here: https://imgur.com/a/ifTSZhH. What should I be looking for in terms of adherence to the codec?

**Comment by robertsonics** [comment]
Ideally, you could isolate a single file that causes the problem. What happens when you loop each of the "bad" 4 individually? If any one file does it, I'd love to get that file and see if I can duplicate the issue.

Are you making sure to save your wave files without any metadata?

**Comment by djrobby** [comment]
Yes, I will most definitely send you the file(s) tomorrow once I’ve isolated the issue. As for metadata, yes. I used SoundForge to downsample 48kHz/24-bit stereo wav files to 44.1kHz/16-bit mono and stripped away any fatty metadata such as title, author, genre etc. I believe the metadata I sent you the screenshots of is the inherent filetype info that goes along with every wav file. I will also double-check that fact tomorrow to be 200% certain.

**Comment by qniens** [comment]
We have been using the Tsunami a lot in our projects. In our experience clicking sounds/bad playback can be caused by a SD-Card that gets corrupted/damaged. The SD-Card will still work on your PC but in the Tsunami they will produce unusable audio.

The 'destruction' of the SD-card happens usually after a endurance test of +- 24 hours.
Solution was to which to Industrial grade SD-cards or the 'Endurance max' series of Sandisk.

**Comment by djrobby** [comment]
@robertsonics I apologize for the delay in my response but something weird happened (on my end) the day after I reported the issue and I wanted to investigate it further prior to responding. So, it looks like after I let the completely different tracks play for a day or so, the problematic tracks (track 1-4) started working just fine. Since then, I have not been able to replicate the issue. This is knowing that no audio files changed and nor did my code base they drives them (via serial). I could have something to so with the SD card itself as @qniens mentioned. I'm using `SanDisk Ultra 32GB A1 microSDHC` formatted with `FAT32` filesystem having a `32KB Allocation Unit Size`. As per @qniens, `SanDisk Max Endurance 32GB V30 microSDHC` might be a better option. Please let me know your thoughts.

Also, do you know which MCU you plan on moving to for the 10-channel option and how many layers the PCB will be? The reason I ask is becuase I might have to have our engineer re-design the layout to fit inside a metal enclosure that provides industrial Phoenix Connectors as field termination points for our application (along with a physical serialport which might end up being in the form of a thru-hole RJ45 connector).

**Comment by robertsonics** [comment]
There are so many microSD cards available and they are changing all the time, so there's no way I can make a statement about any cards I haven't personally tried. That said, more demanding customers who have used industrial grade cards generally report that they perform well.

A new player will likely be based on the STM32H750. Not completely sure of your intentions, but just keep in mind that the firmware is not open source and the downloadable files on my website will not run on custom hardware. I don't want anyone to go to the trouble of designing their own board and then discover my firmware won't run on it.

**Comment by djrobby** [comment]
My intent is NOT to customize any part of the firmware or schematic design. Due to the requirements of my application, instead of soldering components onto the Tsunami board after the fact, I figured it would be easier to do a custom layout and place actual IO components onto the PCB for easy of use and fitting inside the metal enclosure (rendering attached). I hope that's OK?

![SS_20220719_12](https://user-images.githubusercontent.com/22199/179811802-12b44810-d086-42a9-b28b-dac7d9b788c2.jpg)

**Comment by robertsonics** [comment]
I just sent you an email.

**Comment by djrobby** [comment]
Hi Robert, happy new year! I hope you're well. Is there any update on if and when the `Tsunami WAV Trigger` inventories will be replenished? If not, are there any plans to produce a similar board with a different/more readily available MCU? Please let me know. Thank you 🙏

**Comment by robertsonics** [comment]
There are several components that are causing supply problems for Tsunami right now, and SparkFun is actively working on the issue. I'm told there may be a run of either the original or the new "25-voice" version (the CPU has less memory but is more available) in the next few weeks, but both processors remain elusive, so I don't know how many they'll be able to build before running out again.

I couldn't get sufficient (any) quantity of the STM32H750 for my own board design, so I pivoted and have a new player based on the Daisy Seed - see the "Daisy Rogue WAV" on my website. It will be more capable than Tsunami, but will be a different form factor - smaller footprint but much taller to accommodate the Seed.

**Comment by djrobby** [comment]
Thank you Robert. Any ideas about when `Daisy Rogue WAV` module will be ready for purchase? Also, similar to `Tsunami WAV Trigger`, will it support control via RS232?

**Comment by robertsonics** [comment]
Maybe a month? I'm currently working on a board support packege for the Electro-Smith libDaisy codebase so that the Daisy community can use the board to write their own software. As soon as that's done, I'll probably start offering the board as just a multichannel Daisy compatible board, but without firmware. I can only start selling the Robertsonics Player version (board with my firmware) when I feel the firmware is ready. Perhaps 2 months?

And yes, it will support a similar serial control interface. The init file concept (and the Configurator app) has been replaced with csv "preset" files that can be created/edit with Excel or Google Sheets. This is because of the vastly increased MIDI capabilities.

**Comment by baileysage** [comment]
@qniens I'd be interested if you have more information about which SD cards you're using. 

> The 'destruction' of the SD-card happens usually after a endurance test of +- 24 hours. Solution was to which to Industrial grade SD-cards or the 'Endurance max' series of Sandisk.

I am having a similar 'destruction' of SD cards happening on my project since it runs all day in a museum, and I am pretty sure they leave it on 24/7 when it's working. 

I am using the diagnostic firmware @robertsonics provided in his blog to measure the read times. Out of the box, the SanDisk High Endurance cards are questionable, and I'm not even willing to try the SanDisk Industrial cards given the measured read times. I'm running them through a "stress test" along with a Kingston Industrial card (that I could only get one of) currently. Pre-test numbers are below: 

Card | Min Read Time (us) | Max Read Time (us)
-- | -- | --
8 GB   Kingston Industrial | 324 | 578
32 GB SanDisk High Endurance 1 | 440 | 2461
32 GB SanDisk High Endurance 2 | 441 | 3114
8 GB SanDisk Industrial 1 | 354 | 13215
8 GB SanDisk Industrial 2 | 354 | 13064

I am reminding the museum they agreed in the contract to add a switch to the outlet this system is plugged into so it gets turned off along with the lights at the end of every day, but I am hoping to find a more reliable memory solution either way.

I saw the Octophon repo and I'd be happy to "upgrade" them to a RasPi with an SSD if the Octophon hat will be available soon.

**Comment by robertsonics** [comment]
It really depends on the frequency of those >10ms reads and what kind of polyphony you require. If you're only playing a small number of voices, then 13 ms, while not great, is probably not going to cause problems provided it doesn't happen multiple times in a row. I have a customer that uses the SanDisk industrial cards in a commercial product that uses similar firmware and he has not reported any degraded cards. If you're maxing out polyphony however, then yes those will be a problem. What is your typical use case with respect to polyphony?

My 8-channel hat is a dead end, killed by the unavailability of parts and Raspberry Pi's (at the time.) Since then, HiFiBerry has introduced an 8 channel DAC8x board. I haven't tried it. The Octaphon repo code could be pretty easily adapted to any available multichannel hat. One thing to note: The Octaphon code does not stream audio from media - it loads tracks into memory, so you're limited by the available RAM in the RPi (2G, 4G, 8G, etc.)

**Comment by qniens** [comment]
> @qniens I'd be interested if you have more information about which SD cards you're using.
> 
> > The 'destruction' of the SD-card happens usually after a endurance test of +- 24 hours. Solution was to which to Industrial grade SD-cards or the 'Endurance max' series of Sandisk.
> 
> I am having a similar 'destruction' of SD cards happening on my project since it runs all day in a museum, and I am pretty sure they leave it on 24/7 when it's working.
> 
> I am using the diagnostic firmware @robertsonics provided in his blog to measure the read times. Out of the box, the SanDisk High Endurance cards are questionable, and I'm not even willing to try the SanDisk Industrial cards given the measured read times. I'm running them through a "stress test" along with a Kingston Industrial card (that I could only get one of) currently. Pre-test numbers are below:
> 
> Card	Min Read Time (us)	Max Read Time (us)
> 8 GB Kingston Industrial	324	578
> 32 GB SanDisk High Endurance 1	440	2461
> 32 GB SanDisk High Endurance 2	441	3114
> 8 GB SanDisk Industrial 1	354	13215
> 8 GB SanDisk Industrial 2	354	13064
> I am reminding the museum they agreed in the contract to add a switch to the outlet this system is plugged into so it gets turned off along with the lights at the end of every day, but I am hoping to find a more reliable memory solution either way.
> 
> I saw the Octophon repo and I'd be happy to "upgrade" them to a RasPi with an SSD if the Octophon hat will be available soon.

In the end we stopped using the Tsunami because of the unreliable playback. I understand @robertsonics decision for no buffer because of the polyphony use as a instruments, but was giving too many glitches for us. We made our own player that buffers the tracks before playing, That being said, we still use them in some projects, and with the Intenso cards we have the best results:
https://www.intenso.de/en/products/memory-cards/microsd-card-class-10/#produktdetails
3413450 (4Gb) version

**Comment by baileysage** [comment]
We're simulating lake sounds in an odd shaped room. We have five speakers playing layered ambient tracks on loop with birds and frog tracks playing "at random" along with human noises such as boats, a dog, and boots walking down a dock. I have the system tied into occupancy sensors so the ambient noises play quietly all of the time, and then increase in volume and trigger the birds etc when someone enters the room.  

Unfortunately, I'm getting errors with cards showing 4ms read times, so that's why I immediately discounted the 13ms ones. I was only able to get the 25 voice board when I installed the system, though. I now have a couple of the 32 voice ones, so I will test again to see if that changes anything. 

On the positive side, we're only ever going to need about half a GB of recordings at any time. So the RAM on a RasPi won't be a limiting issue if we go the Octophon/HiFiBerry route. 

Bedankt @qniens, I'll see if I can get a few of those Intenso cards to the US to try out while I work on a more permanent solution for this use case. 

I got the Tsunami because I was impressed at all the functionality it offers me. I'm still impressed, given that it got me this far with the abuse I've been submitting it to!

**Comment by robertsonics** [comment]
Not to complicate matters, but we did have a batch of 25-voice Tsunamis that came through with some out of tolerance parts related to the audio codec. These boards were experiencing glitches totally unrelated to the microSD card. SparkFun has worked with affected customers to replace the units. So far, we've only ever seen this issue on a small number of 25-voice boards.

The symptoms of this issue are a crackling sound even when playing a single voice. The codec's PLL stops working and the output is forced to 0 for short periods of time. Has nothing to do with how many voices are playing.

What is the nature of the errors you're experiencing? Can you easily swap in an original 32-voice Tsunami using the same microSD card and see if that changes anything?

**Comment by baileysage** [comment]
One bout of COVID and one holiday later.... 

We were getting our on-site glitch at the same runtime each time it failed. Usually on the second time through the "ambient" sounds. Infrequently on the first time or third time through, but always at the same point in those files, even if I disabled the triggers for playing the additional sounds, so they were the only ones playing. And I could play other files on repeat and not get the issue, so it really does seem like a bad sector issue and not one of the bad 25 voice boards. We also had full-on cover-your-ears-level digital "static", similar to what you get if you yank the SD card while it's playing, not a crackling. 

I haven't run that card on a 32 voice board yet as I was avoiding going to my coworking space while I was contagious, but I can still do so if that's helpful.

Additional reason I don't think it's the 25 voice defect: The one SP3D NAND card I put through a stress test had a sector so bad, I thought your test firmware had hung out of the gate. Luckily, I got distracted by a conversation long enough that the first file finally got fully tested before I gave up on it. The firmware performed normally after the first file was completed, but it had a max read of 95,889ms and a good chunk of reads >20ms... which explains why I thought it had stalled out.

Other news since my last update, the 8GB Kingston Industrial card mentioned above tested with the _exact same_ min and max read times after running for 17 days straight. I did get some of the Intenso cards ordered. (Bonus: they're CHEAP!!!) But I will be installing a Kingston card at the museum while I wait on them to show up and while I work on a RAM based option. 

I have a second project with this museum to complete yet. It only plays sounds in response to button presses and proximity sensors. It sounds like the Tsunami will work just fine there!

**Comment by qniens** [comment]
The Intenso cards are indeed cheap. In our test and usage the have a constant low random acces time.

**Comment by robertsonics** [comment]
If you don't mind my asking, where are you getting the Intenso cards inexpensively? I just ordered a couple to test on Amazon, US, and they were definitely not cheap.

I'm working on the new WAV Trigger Pro and have built the microSD card test into the regular firmware. It also constantly monitors the read times under normal use and will alert the user (with the LED) if any single read takes longer than 20ms.

FYI, my polyphony test pushes the board to 24 stereo voices all pitched up at 700 cents, and I've been having good results with the (white) SanDisk Industrial 8GB cards.

**Comment by baileysage** [comment]
I picked up the Intenso cards on grooves.land. They shipped through the Netherlands and if the tracking is accurate, looks like it'll be 3 weeks from order to arrival.
